porch/pkg/cache/repository.go

```diff
@@ -16,58 +16,44 @@ package cache
 
 import (
 	"context"
-	"sort"
-	"strings"
 	"sync"
 	"time"
 
 	"github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
 	"github.com/GoogleContainerTools/kpt/porch/pkg/repository"
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/trace"
-	"golang.org/x/mod/semver"
 	"k8s.io/klog/v2"
 )
 
 var tracer = otel.Tracer("cache")
 
+// We take advantage of the cache having a global view of all the packages
+// in a repository and compute the latest package revision in the cache
+// rather than add another level of caching in the repositories themselves.
+// This also reuses the revision comparison code and ensures same behavior
+// between Git and OCI.
+
+var _ repository.Repository = &cachedRepository{}
+var _ repository.FunctionRepository = &cachedRepository{}
+
 type cachedRepository struct {
 	id     string
 	repo   repository.Repository
 	cancel context.CancelFunc
 
-	mutex          sync.Mutex
-	cachedPackages []*cachedPackageRevision
+	mutex                  sync.Mutex
+	cachedPackageRevisions []*cachedPackageRevision
+	cachedPackages         []*cachedPackage
+
 	// TODO: Currently we support repositories with homogenous content (only packages xor functions). Model this more optimally?
 	cachedFunctions []repository.Function
 	// Error encountered on repository refresh by the refresh goroutine.
 	// This is returned back by the cache to the background goroutine when it calls periodicall to resync repositories.
-	refreshError error
+	refreshRevisionsError error
+	refreshPkgsError      error
```

Comment:
I separated this into two because I felt that an error with getting packages didn't need to block getting package revisions.